### PR TITLE
Incubating feature: expression evaluation

### PR DESCRIPTION
**Pull request by phensley**
This branch includes a new instruction EVAL which evaluates expressions with JavaScript-like semantics. The plan is to release this in the TypeScript compiler as an incubating feature for evaluation by various teams, collect feedback and produce the final form of the instruction.

A simple example with immediate output of the result:

```
{.eval 3 * 4 + 9 / 2}
{.eval 3 * (4 + 9) / 2}
{.eval 3 * (4 + (9 / 2))}
{.eval (3 * 4 + 9) / 2}
```

Would output:

```
16.5
19.5
25.5
10.5
```

The test cases `math.test.ts` and `engine.test.ts` include more usage examples.

The commit message follows with details:

    Expression evaluation using an extended version of Dijkstra's "shunting
    yard" algorithm with JavaScript semantics.  Dijkstra's algorithm was
    chosen as it is simple, sufficient, and can be implemented compactly,
    minimizing the size of the code and making it easier to verify correct.

    Features:

     - Number, Boolean, Null, and String types
     - String parsing of backslash char, hex, and unicode escapes
       - "\n", "\x20", "\u2018", "\u0001f600" are all valid strings
     - Decimal and hex numbers
     - Assignment statements
     - Variable references
     - Operator precedence
     - Common math, binary and logical operators
     - Both strict and non-strict equality
     - Nesting of sub-expressions with parenthesis
     - Single pass evaluation of multiple semicolon-delimited expressions
     - String equality and concatenation
     - Function calls with a variable number of arguments
       - example: max(1, 2, 3, ..., N) is accepted
     - Type conversion functions
     - Common constants (pi, e)

    The algorithm assumes the expression is well-formed and performs minimal
    validation.  A malformed expression should either terminate evaluation
    early or produce no value.

### CODE REVIEW

**Review by mattlyons0 — APPROVED** [review]
This is a big change, but nothing sticks out to me that looks wrong about it. It seems very well thought out and cleanly written. The comments are very useful and tests are a nice touch.

The only potential concern that comes to mind is exception handling in the mathematic operations (things like dividing by 0, parsing things that potentially fail parseint...) which may throw, but I'm assuming this is something that is probably already handled within the library at a higher level. Plus this is the frontend renderer and a new feature so it is much less of a concern for the initial PR.

**Comment by phensley** [comment]
Thanks @mattlyons0  - the cases you outlined (`parseInt()`, `1 / 0`, `Number()`, etc) in JavaScript will return `NaN` for invalid input, which should be safe for evaluation. The tokenizer accept criteria for numbers are strictly defined so it should not accept anything that would throw an error.

The code coverage for the expression engine is currently at 100%, but let me know if you think of cases that might raise errors. It's easy to wrap this in a try/catch if needed for anything unexpected.

File: src/matcher.ts [hunk 1]
```diff
@@ -139,7 +183,8 @@ export class Matcher {
   /**
    * Match one or more variable bindings.
    */
-  matchBindings(): any { // TODO:
+  matchBindings(): any {
+    // TODO:
```
Comment (brianjcarroll) [inline src/matcher.ts]: Is there supposed to be a todo here?

**Comment by mattlyons0** [comment]
Sounds good to me, thanks Pat! NaN what a beautiful thing (in this case lol)

**Review by brianjcarroll — APPROVED** [review]

**Comment by brianjcarroll** [comment]
Agreeing with @mattlyons0 here - PR looks great, nothing that stands out to me

File: src/matcher.ts (hunk 1 above)
Comment (phensley) [reply src/matcher.ts]: Looks like an old comment that got repositioned when the file was reformatted. I'll clean it up when I do the final merge!

**Comment by phensley** [comment]
My goal is to merge this on Friday morning, so add your reviews before then, or mark resolved -- thx

**Comment by phensley** [comment]
Also if you have concerns and need more time to review let me know

**Comment by jshields-squarespace** [comment]
I'm not very familiar with the TypeScript or Java implementations, so I'll just ask this question as a sanity check:

Do you foresee any issues with implementing the exact same behavior in Java, particularly when it comes to the mathematical edge cases that Matt pointed out like `1 / 0 = Infinity`?

I've personally encountered some issues stemming from the fact that JavaScript `Number`s are all floating point, while there's a tendency in Java to favor integer data types when possible. But perhaps if the Java implementation treats all numbers as floating points then the behavior will be the same?

**Comment by phensley** [comment]
The semantics of the expression engine are essentially equivalent to those of JavaScript. This simplifies the implementation of the TypeScript compiler: implementing the expression language's behaviors directly in terms of JavaScript expressions (where those behaviors are well-defined). The Java port is underway, and the various operations, representations, and type conversions will need to be adapted to produce equivalent behavior. I don't currently foresee any complications with achieving full compatibility.

The behavior of 64-bit floating point operations on both Java's `double` and JavaScript's `Number` type are essentially equivalent since both languages implement IEEE 754.  However there are several areas where they clearly differ and these cases will need to be handled explicitly in the Java compiler. For example, JavaScript's bitwise operators treat `Number` values as 32-bit integer internally whereas Java's bitwise operators can work with 64-bit `long` values.  Also, in JavaScript `1 / 0 == Infinity` and `0 / 0 == NaN`, where Java raises `ArithmeticException`, etc.

There are also differences in how numbers are formatted as strings. Java has a variety of ways of formatting numbers, while JavaScript's string representations may be implementation-dependent, e.g. whether a number is displayed as decimal or exponential, and the default number of decimal digits displayed for floating point, may differ between browsers / engines.  I'll be explicitly choosing the string representation, emitting either decimal or exponential format based on the magnitude of the number, and this will be implemented identically for both compilers.

Given all of this, the test coverage will need to ensure the two compilers produce identical output for a variety of edge cases. This is one reason the feature is "incubating" -- it will exist in the production compiler but be disabled until the feature is deemed stable, safe, and compatible.

**Comment by phensley** [comment]
Thanks -- I'm going to merge this so it can be made available on the internal docs / playground site for experimentation. Next phase will be to complete the Java implementation and tweak things to reach compatibility. Feel free to open tickets or ping me with questions!
